metrics/factory_test.go

```diff
@@ -0,0 +1,43 @@
+package metrics
+
+import (
+	"testing"
+	"time"
+
+	"context"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewPublisherWithDSN(t *testing.T) {
+	assert := assert.New(t)
+
+	var cancelFuncs []context.CancelFunc
+	for _, testCase := range []struct {
+		DSN     string
+		isValid bool
+	}{
+		{"http://%41:8080/", false},
+		{"random://", false},
+		{"discard://", true},
+		{"datadog://", false},
+		{"datadog://?namespace=my_namespace", true},
+		{"datadog://?namespace=my_namespace&gostats=false", true},
+	} {
+		if testCase.isValid {
+			publisher, runner := NewMetricsRunnerFromDSN(testCase.DSN)
+			assert.NotNil(publisher)
+			assert.NotNil(runner)
+			ctx, cancel := context.WithCancel(context.Background())
+			cancelFuncs = append(cancelFuncs, cancel)
+			go runner.Run(ctx)
```

Comment:
I removed that and still works, but I think It's good to see that they run can its context can be cancelled, so I will change the test name to refer to that.